File: Web/Controllers/SubjectController.cs
using System.Collections.Generic;
using System.Data.Entity;
using System.Linq;
using System.Web.Mvc;
using Web.Data;
using Web.Models;

namespace Web.Controllers
{
    public class SubjectController : Controller
    {
        private readonly AppDbContext context = new AppDbContext();
        public ActionResult Index()
        {
            return View();
        }

        [HttpGet]
        public ActionResult GetSubjects()
        {
            List<Subject> subjects = context.Subjects.ToList();

            subjects.ForEach((subject) =>
            {
                subject.Teacher = context.Teachers.FirstOrDefault(t => t.Id == subject.TeacherId);
                subject.Course = context.Courses.FirstOrDefault(c => c.Id == subject.CourseId);

                List<Student> students = new List<Student>();

                var subjectStudents = context.SubjectStudents.Where(ss => ss.SubjectId == subject.Id).ToList();
                subjectStudents.ForEach((subjectStudent) =>
                {
                    students.Add(context.Students.FirstOrDefault(s => s.Id == subjectStudent.Id));
                });

                subject.Students = students;
            });

            return Json(subjects, JsonRequestBehavior.AllowGet);
        }

        [HttpPost]
        public ActionResult Insert(Subject subject)
        {
            context.Subjects.Add(subject);
            context.SaveChanges();

            return RedirectToAction("Index");
        }

        [HttpPut]
        public ActionResult Edit(Subject subject)
        {
            var entry = context.Entry(subject);
            context.Set<Subject>().Attach(subject);
            entry.State = EntityState.Modified;

            context.SaveChanges();

            return new HttpStatusCodeResult(200);
        }

        [HttpDelete]
        public ActionResult Delete(int id)
        {
            Subject subject = context.Subjects.FirstOrDefault(s => s.Id == id);

            context.Subjects.Remove(subject);
            context.SaveChanges();

            return new HttpStatusCodeResult(200);
        }
    }
}

File: Web/NgControllers/subjectController.js
app.controller("SubjectController", function ($scope, $http) {
    $scope.insertedSubject = {};
    $scope.currentId = null;
    $scope.isEdit = false;

    $scope.getValues = function (route, type) {
        $http.get(route)
            .then((res) => {
                switch (type) {
                    case "Subject":
                        $scope.Subjects = res.data;
                        break;
                    case "Teacher":
                        $scope.Teachers = res.data;
                        break;
                    case "Course":
                        $scope.Courses = JSON.parse(res.data);
                        break;
                }

            })
    };

    $scope.insertOrEditSubject = function () {
        if ($scope.insertedSubject && $scope.insertedSubject.SubjectName && !$scope.isEdit) {
            console.log($scope.insertedSubject);
            $http.post("/Subject/Insert", $scope.insertedSubject)
                .then(() => {
                    $('#editModal').modal('hide');

                    $scope.clearEditValues();

                    $scope.getValues('/Subject/GetSubjects', 'Subject');
                });
        } else if ($scope.insertedSubject && $scope.insertedSubject.SubjectName && $scope.isEdit) {
            $http.put("/Subject/Edit", $scope.insertedSubject)
                .then(() => {
                    $('#editModal').modal('hide');

                    $scope.clearEditValues();

                    $scope.getValues('/Subject/GetSubjects', 'Subject');
                });
        }
    };

    $scope.deleteSubject = function () {
        if ($scope.currentId) {
            $http.delete("/Subject/Delete?id=" + $scope.currentId)
                .then(() => {
                    $('#deleteModal').modal('hide');

                    $scope.currentId = null;

                    $scope.getValues('/Subject/GetSubjects', 'Subject');
                });
        }
    };

    $scope.clearEditValues = function () {
        $scope.isEdit = false;
        $scope.insertedSubject = {}
    };

    $scope.setCurrentId = function (currentId) {
        $scope.currentId = currentId;
    };

    $scope.setSubjectToEdit = function (subject) {
        $scope.insertedSubject.Id = subject.Id;
        $scope.insertedSubject.CourseId = subject.CourseId;
        $scope.insertedSubject.TeacherId = subject.TeacherId;
        $scope.insertedSubject.SubjectName = subject.SubjectName;

        console.log($scope.insertedSubject);

        $scope.isEdit = true;
    };
})